cdc/sink/manager.go

```diff
@@ -0,0 +1,267 @@
+	// do nothing
+	return nil
+}
+
+func (t *tableSink) EmitRowChangedEvents(ctx context.Context, rows ...*model.RowChangedEvent) error {
+	t.buffer = append(t.buffer, rows...)
+	return nil
+}
+
+func (t *tableSink) EmitDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	// the table sink doesn't receive the DDL event
+	return nil
+}
+
+func (t *tableSink) FlushRowChangedEvents(ctx context.Context, resolvedTs uint64) (uint64, error) {
+	i := sort.Search(len(t.buffer), func(i int) bool {
+		return t.buffer[i].CommitTs > resolvedTs
+	})
+	if i == 0 {
+		atomic.StoreUint64(&t.emittedTs, resolvedTs)
+		return t.manager.flushBackendSink(ctx)
+	}
+	resolvedRows := t.buffer[:i]
+	t.buffer = t.buffer[i:]
+
+	err := t.manager.backendSink.EmitRowChangedEvents(ctx, resolvedRows...)
+	if err != nil {
+		return t.manager.getCheckpointTs(), errors.Trace(err)
+	}
+	atomic.StoreUint64(&t.emittedTs, resolvedTs)
+	return t.manager.flushBackendSink(ctx)
+}
+
+func (t *tableSink) getEmittedTs() uint64 {
+	return atomic.LoadUint64(&t.emittedTs)
+}
+
+func (t *tableSink) EmitCheckpointTs(ctx context.Context, ts uint64) error {
+	// the table sink doesn't receive the checkpoint event
+	return nil
+}
+
+func (t *tableSink) Close() error {
+	t.manager.destroyTableSink(t.tableID)
+	return nil
+}
+
+type bufferSink struct {
+	Sink
+	buffer chan struct {
+		rows       []*model.RowChangedEvent
+		resolved   bool
+		resolvedTs model.Ts
+	}
+	checkpointTs uint64
+}
+
+func newBufferSink(ctx context.Context, backendSink Sink, errCh chan error, checkpointTs model.Ts) Sink {
+	sink := &bufferSink{
+		Sink: backendSink,
+		buffer: make(chan struct {
+			rows       []*model.RowChangedEvent
+			resolved   bool
+			resolvedTs model.Ts
+		}, defaultBufferChanSize),
+		checkpointTs: checkpointTs,
+	}
+	go sink.run(ctx, errCh)
+	return sink
+}
+
+func (b *bufferSink) run(ctx context.Context, errCh chan error) {
+	changefeedID := util.ChangefeedIDFromCtx(ctx)
+	advertiseAddr := util.CaptureAddrFromCtx(ctx)
+	metricFlushDuration := flushRowChangedDuration.WithLabelValues(advertiseAddr, changefeedID)
+	metricBufferSize := bufferChanSizeGauge.WithLabelValues(advertiseAddr, changefeedID)
+	for {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			if err != nil && errors.Cause(err) != context.Canceled {
+				errCh <- err
+			}
+			return
+		case e := <-b.buffer:
+			if e.resolved {
+				start := time.Now()
+				checkpointTs, err := b.Sink.FlushRowChangedEvents(ctx, e.resolvedTs)
+				if err != nil {
+					if errors.Cause(err) != context.Canceled {
+						errCh <- err
+					}
+					return
+				}
+				atomic.StoreUint64(&b.checkpointTs, checkpointTs)
+
+				dur := time.Since(start)
+				metricFlushDuration.Observe(dur.Seconds())
+				if dur > 3*time.Second {
+					log.Warn("flush row changed events too slow",
+						zap.Duration("duration", dur), util.ZapFieldChangefeed(ctx))
+				}
+				continue
+			}
+			err := b.Sink.EmitRowChangedEvents(ctx, e.rows...)
```

Comment:
Do we need to monitor the duration of `EmitRowChangedEvents` too? If so, may be we can resue the metric `flushRowChangedDuration` by adding one more label